EIPS/eip-6997.md

```diff
@@ -0,0 +1,371 @@
+---
+eip: 6997
+title: ERC-721 with transaction validation step.
+description: A new validation step for transfer and approve calls, achieving a security step in case of stolen wallet.
+author: eduardfina (@eduardfina), Eduard López i Fina
+discussions-to: https://ethereum-magicians.org/t/erc721-with-a-validation-step/14071
+status: Draft
+type: Standards Track
+category: ERC
+created: 2023-05-07
+requires: 721
+---
+
+## Abstract
+
+This standard is an extension of [ERC-721](./eip-721.md). It defines a new validation functionality to avoid wallet draining, every `transfer` or `approve` will be locked waiting for validation.
+
+## Motivation
+
+The power of the Blockchain is at the same time its weakness, giving the user full responsibility for their data.
+
+Many cases of NFT theft currently exist, and current NFT anti-theft schemes, such as transferring NFTs to cold wallets, make NFTs inconvenient to be used.
+
+Having a validation step before every `transfer` and `approve` would bring the Smart Contract developers the possibility to create a security NFT anti-theft schemes.
+
+When I was developing this standard I was thinking of a system where there would be a validator address that would be responsible for validating all Smart Contract transactions.
+
+This address would be connected to a dApp where the user could see the validation requests of his NFTs and accept the correct ones.
+
+Giving this address only the power to validate transactions would make a much more secure system where to steal an NFT the thief would have to have both the user's address and the validator address simultaneously.
+
+## Specification
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+[ERC-721](./eip-721.md) compliant contracts MAY implement this EIP.
+
+When `transfer` is called by the owner, the transfer MUST be frozen and a `TransferValidation` MUST be created.
```

Comment:
What does "frozen" mean in this context? Does `transfer` revert, or return successfully?

Reply:
As you can see at the example implementation, the ERC-721 `_transfer` function is override. 
Instead of transferring the token it creates a validateTransfer object.
Then with the function _validateTransfer is when the ERC-721 `_transfer` function is called.
The same happen with `_approve` and `_setApprovalForAll `

This step of waiting for validation is what I call "frozen".

Reply:
> As you can see at the example implementation

We prefer if the standard is implementable without having to refer to the reference implementation. All of the externally visible behaviour should be defined within the specification section. You could add something like:

```markdown
A "frozen" operation, like a `transfer`/`safeTransferFrom`/`approve`/etc. SHALL
return <!-- insert text here --> and SHALL NOT change the state of the contract
aside from <!-- insert text here -->.
```

This is just an example. Please write it as appropriate for your proposal, and include it in your document.

> the ERC-721 `_transfer` function is override.

Ah, okay. The Open Zeppelin `_transfer`, `_approve`, and similar functions aren't actually part of the ERC-721 specification. They are an implementation detail. In other words, some implementations of ERC-721 don't actually have those functions. For example, [solmate's ERC-721 implementation](https://github.com/transmissions11/solmate/blob/main/src/tokens/ERC721.sol) doesn't have any of them.

Reply:
Thank you, I changed the specification 👍